stages/org.osbuild.ostree.aleph

```diff
@@ -0,0 +1,165 @@
+
+
+def aleph_container(tree, imgref):
+    # extract the image name from the imgref
+    imgref_list = imgref.split(':')
+    if imgref_list[0] in ["ostree-remote-registry", "ostree-remote-image"]:
+        img_name = ':'.join(imgref_list[2:])
+    elif imgref_list[0] in ["ostree-image-signed", "ostree-unverified-registry"]:
+        img_name = ':'.join(imgref_list[1:])
+    else:
+        raise ValueError(f"Image ref {imgref} has unsupported type (supported: 'ostree-remote-registry', \
+                         'ostree-remote-image', 'ostree-image-signed', or 'ostree-unverified-registry')")
+
+    img_name = img_name.removeprefix('docker://')
+
+    extra_args = []
+    extra_args.append(f"--repo={tree}/ostree/repo")
+    extra_args.append(f"registry:{img_name}")
+
+    container_data_json = ostree.cli("container", "image", "metadata", *extra_args).stdout.rstrip()
+    container_data = json.loads(container_data_json)
+
+    extra_args.append("--config")
+    container_data_config_json = ostree.cli("container", "image", "metadata", *extra_args).stdout.rstrip()
+    container_data_config = json.loads(container_data_config_json)
+
+    aleph_digest = container_data['config']['digest']
+    aleph_ref = f"docker://{imgref}"
+    aleph_version = container_data_config['config']['Labels']['org.opencontainers.image.version']
+    aleph_container_image = container_data_config['config']['Labels']
+
+    aleph_version_data = {
+        "osbuild-version": osbuild.__version__,
+        "ref": aleph_ref,
+        "version": aleph_version,
+        "container-image": {
+            "image-name": img_name,
+            "image-digest": aleph_digest,
+            "image-labels": aleph_container_image
+        }
+    }
+
+    # the 'ostree.commit' label will be optional in the future so
+    # prevent hard failing if key is not found
+    aleph_ostree_commit = container_data_config['config']['Labels'].get('ostree.commit')
+    if aleph_ostree_commit is not None:
+        aleph_version_data["ostree-commit"] = aleph_ostree_commit
+
+    return aleph_version_data
+
+
+def construct_aleph_json(tree, origin):
+    deploy_type, imgref = ostree.parse_origin(origin)
+    data = {}
+    # null deploy_type and imgref error is caught in the parse_origin() function
+    if deploy_type == "container":
+        data = aleph_container(tree, imgref)
+    elif deploy_type == "ostree_commit":
+        data = aleph_commit(tree, imgref)
+    else:
+        raise ValueError("Unknown deployment type")
+
+    return data
+
+
+def main(tree, options):
+    coreos_compat = options.get("coreos_compat", False)
+    dep = options.get("deployment", None)
+    osname = ""
+    ref = ""
+    serial = ""
+    origin = ""
+
+    # if deployment is specified, use it to find the origin file.
+    # otherwise, autodetect the only deployment found in the tree.
+    if dep is not None:
+        osname = dep.get("osname", "")
+        ref = dep.get("ref", "")
+        serial = dep.get("serial", "0")
```

Comment:
Is there ever a situation where this information isn't available and needs to be detected at build time instead of when generating the manifest?  This might be okay, depending on the scenario.  The lack of explicitness is a little concerning.  If it's for convenience, then we definitely shouldn't do it.  If it's because information isn't accessible during manifest generation, then that's something we can examine.

(this is a comment about the whole idea of detecting the deployment)

Reply:
I can't really conceive of a good reason to create a disk image with two deployments, so IMO the default should be to just not require the deployment information to be specified and have it autodetected, while also allowing it to be specified if that should ever be needed.

Reply:
I am not aware of a strong motivation to have more than one initial deployment.  Requiring pipelines to thread all this stuff through just seems like overhead.

If there ever were, it would make way more sense to me to have something like an optional `deployment_index` - that's how ostree internally (and externally) operates currently.

Reply:
Sorry for the confusion, I realise now that it looked like my comment was about the deployment index/serial only. 
 My comment was about the osname and ref in particular (the deployment index defaulting to 0 is probably fine).

> Requiring pipelines to thread all this stuff through just seems like overhead.

That's the job of manifest generators though.  Stages should have all the information they need without needing to "sniff" the filesystem, unless it's absolutely necessary.  The manifest should be as explicit as possible and all the information about the build should be visible in the pipelines.
The aleph stage is basically writing some formatted metadata.  The stage should have all the information needed to write this and that information should be visible in the stage options and any dependencies on the state of the pipeline tree should be minimal.

Reply:
If there can only be one answer to a question then why make the manifest require it?

One example is the `bootupd` stage I'm adding later (WIP over in https://github.com/dustymabe/osbuild/pull/11). I'm actually proposing for that one we don't even start with adding these options to the stage. @lukewarmtemp worked on this stage, otherwise I would have probably done the same here for the aleph stage.

Reply:
I wrote this in chat already but I'll paste it here for visibility:
Having everything be explicit in stages is a feature and an important one.
It makes it testable: I can run the stage and verify its output without needing to run it against an ostree deployment.
It makes it predictable: I know what the file will look like just by looking at the stage on its own.
And it makes it more flexible: We might have a reason to run it before the deployment, or run it in a separate pipeline and copy everything into its place in a separate one.  These are all things that we sometimes end up doing for various reasons with other stages.
And yeah we can bend on this requirement for some stages sometimes when needed, but I don't know if this is one of those cases.

Reply:
I dropped the commit that modified the deployment_path function update.

Reply:
> I've been wondering about that. With everything we're working on in osbuild to leverage everything bootc does, wouldn't it be more effective if we were all working on doing it all (FCOS included) with bootc or at least in a way compatible with what we're doing there?

I think the bigger limiting factor here is RHCOS where we're trying to ship something that will depend on this way sooner than we'll ever get to bootc.